src/ogdf/basic/GraphCopy.cpp

```diff
@@ -130,57 +131,88 @@ void GraphCopy::setOriginalGraph(const Graph* G) {
 	m_eIterator.init(this, nullptr);
 }
 
-void GraphCopySimple::setOriginalEmbedding() {
-	std::vector<adjEntry> newAdjOrder;
-	for (node v : getOriginalGraph()->nodes) {
-		node u = copy(v);
-		if (u == nullptr) {
-			continue;
-		}
-		newAdjOrder.clear();
-		newAdjOrder.reserve(v->degree());
+void copyEmbedding(const Graph& from, Graph& to, std::function<adjEntry(adjEntry)> adjMapFromTo) {
+	AdjEntrySet order(to);
+	for (node fn : from.nodes) {
+		node tn = nullptr;
+		order.clear();
+
 		// add from original according to their order
-		for (adjEntry adj : v->adjEntries) {
-			edge e = copy(adj->theEdge());
-			if (e != nullptr && e->isIncident(u)) {
-				newAdjOrder.push_back(e->getAdj(u));
+		for (adjEntry fadj : fn->adjEntries) {
+			adjEntry tadj = adjMapFromTo(fadj);
+			if (tadj == nullptr) {
+				continue;
 			}
+			if (tn == nullptr) {
+				tn = tadj->theNode();
+			} else {
+				OGDF_ASSERT(tadj->theNode() == tn);
+			}
+			order.insert(tadj);
 		}
+
 		// add remaining dummy edges to the end, also retaining their order
-		for (adjEntry adj : u->adjEntries) {
-			if (isDummy(adj) || !copy(original(adj->theEdge()))->isIncident(u)) {
-				newAdjOrder.push_back(adj);
+		if (tn == nullptr) {
+			continue;
+		}
+		for (adjEntry tadj : tn->adjEntries) {
+			if (!order.isMember(tadj)) {
+				order.insert(tadj);
 			}
 		}
-		OGDF_ASSERT(newAdjOrder.size() == size_t(u->degree()));
-		sort(u, newAdjOrder);
+
+		OGDF_ASSERT(order.size() == size_t(tn->degree()));
+		to.sort(tn, order);
 	}
 }
 
-void GraphCopy::setOriginalEmbedding() {
-	std::vector<adjEntry> newAdjOrder;
-	for (node v : getOriginalGraph()->nodes) {
-		node u = copy(v);
-		if (u == nullptr) {
-			continue;
-		}
-		newAdjOrder.clear();
-		newAdjOrder.reserve(v->degree());
-		// add from original according to their order
-		for (adjEntry adj : v->adjEntries) {
-			if (copy(adj) != nullptr) {
-				newAdjOrder.push_back(copy(adj));
+void GraphCopySimple::copyEmbeddingToOriginal(Graph& orig) const {
+	OGDF_ASSERT(getOriginalGraph() == &orig);
+	copyEmbedding(*this, orig, [this](adjEntry adj) -> adjEntry {
+		node v = original(adj->theNode());
```

Comment:
Excatly! Here we want to be slightly smarter regarding reversed edges, so that `copyEmbedding` even works when that happened, but moving this functionality to the general `copy` could easily break other code implicitly relying on the fact that this method was so far oblivious to reversed edges.